### PR TITLE
Faster name visitor subset

### PR DESCRIPTION
**Pull request by behdad**
https://github.com/fonttools/fonttools/issues/3213#issuecomment-1644158826

### CODE REVIEW

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 1]
```diff
@@ -242,6 +242,9 @@ def removeUnusedNames(ttFont):
             if record.nameID not in visitor.seen:
                 toDelete.add(record.nameID)
 
+        if not toDelete:
+            return
+        log.info(f"Deleting name records with NameIDs {toDelete}")
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py:247]: these three lines above we don't want regardless

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 2]
```diff
@@ -1208,3 +1211,15 @@ def visit(visitor, obj):
...
         visitor.seen.update(obj.paletteLabels)
         visitor.seen.update(obj.paletteEntryLabels)
+
+
+@NameRecordVisitor.register(
+    (
+        ttLib.getTableClass("cmap"),
+        ttLib.getTableClass("glyf"),
+        ttLib.getTableClass("CFF "),
+        ttLib.getTableClass("CFF2"),
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py:1221]: hm I don't know.. what if a font has a huge/deep table that you don't know about? It feels to me the selection of tables to visit should be opt-in, not opt-out

Comment (behdad) [reply Lib/fontTools/ttLib/tables/_n_a_m_e.py:1221]: Sounds good. Let me see if I can implement that easily.

File: Lib/fontTools/subset/__init__.py [hunk 3]
```diff
@@ -2915,32 +2901,10 @@ def prune_pre_subset(self, font, options):
...
-        nameIDs.update(
-            [axis_rec.AxisNameID for axis_rec in stat.table.DesignAxisRecord.Axis]
-        )
-    cpal = font.get("CPAL")
-    if cpal and cpal.version == 1:
-        nameIDs.update(cpal.paletteLabels)
-        nameIDs.update(cpal.paletteEntryLabels)
+def prune_post_subset(self, font, options):
+    visitor = NameRecordVisitor()
+    visitor.visit(font)
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:2906]: why do we need to "visit" the entire font if we know already where we expect to find nameIDs?

Comment (behdad) [reply Lib/fontTools/subset/__init__.py:2906]: PTAL

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 4]
```diff
@@ -1208,3 +1208,16 @@ def visit(visitor, obj):
...
         visitor.seen.update(obj.paletteEntryLabels)
+
+
+@NameRecordVisitor.register(ttLib.TTFont)
+def visit(visitor, font, *args, **kwargs):
+    if hasattr(visitor, "font"):
+        return False
+
+    visitor.font = font
+    for tag in ("GSUB", "GPOS", "fvar", "CPAL", "STAT"):
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py:1219]: how about storing this list of table tags that can contain nameID references as a class attribute of NameRecordVisitor?

Comment (behdad) [reply Lib/fontTools/ttLib/tables/_n_a_m_e.py:1219]: Okay.

Comment (behdad) [reply Lib/fontTools/ttLib/tables/_n_a_m_e.py]: Done.

**Comment by anthrotype** [comment]
it would be nice if @imran-iq could give this PR branch a try with their go-noto font build to confirm it is fast enough?

**Review by anthrotype — APPROVED** [review]

**Comment by behdad** [comment]
Here's a timing I observed:
```
$ time ./fonttools subset --gids=0-10000 --font-number=0 SourceHanNotoCJK.ttc --timing
Took 0.000s to load font
Took 0.000s to compile glyph list
Took 0.000s to load 'maxp'
Took 0.000s to prune 'maxp'
Took 0.041s to load 'cmap'
Took 0.000s to prune 'cmap'
Took 0.000s to load 'post'
Took 0.000s to prune 'post'
Took 0.000s to load 'CFF '
Took 0.017s to prune 'CFF '
Took 0.000s to load 'GPOS'
Took 0.009s to prune 'GPOS'
Took 0.000s to load 'GSUB'
Took 0.015s to prune 'GSUB'
Took 0.012s to close glyph list over 'cmap'
Took 0.068s to close glyph list over 'GSUB'
Took 1.175s to close glyph list over 'CFF '
Took 0.011s to subset 'cmap'
Took 0.000s to subset 'post'
Took 0.013s to subset 'CFF '
Took 0.000s to subset 'VORG'
Took 0.056s to subset 'hmtx'
Took 0.016s to subset 'vmtx'
Took 0.000s to subset 'GDEF'
Took 0.090s to subset 'GPOS'
Took 0.010s to subset 'GSUB'
Took 0.000s to subset GlyphOrder
Took 0.000s to prune 'head'
Took 2.202s to prune 'CFF '
Took 0.000s to prune 'GDEF'
Took 0.002s to prune 'GPOS'
Took 0.005s to prune 'GSUB'
Took 0.186s to prune 'name'
Took 1.160s to compile and save font
Took 5.102s to make one with everything (TOTAL TIME)

real	0m5.222s
user	0m5.147s
sys	0m0.059s
```
Without name closure it takes `4.925` (as in without this PR). The name pruning took `0.186s`, which I think is acceptable.

With the previous approach this was taking minutes before I interrupted it...

**Pull request merged**